### PR TITLE
add python script for ubuntu

### PR DESCRIPTION
**Pull request by S-VIN**
Please write your comments and remarks. I will add comments in code after I correct them.
Python script only for ubuntu, because I download esptool using apt.

### CODE REVIEW

File: flash.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                raise Exception("Wrong input number")
+        except:
+            print("Print a correct number.")
+    return num
+    
+def getCOMPortList():
+    ports = os.listdir("/dev/")
+    output = []
+    for i in range(len(ports) - 1):
+        if((re.search("ttyUSB*", ports[i]) != None) or (re.search("ttyACM*", ports[i]) != None)):
```
Comment (AlexErAmp) [inline flash.py]: According to pep8, the string should not exceed 79 characters. Format your code with python black with `--line-length= 79` option.

File: flash.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+import os
+import subprocess
+import re
+
+def downloadESPtool():
+    output = subprocess.run(["sudo", "apt", "install", "esptool"], text=True)
+    if(output.returncode != 0):
+        raise Exeption(output.returncode)
```
Comment (AlexErAmp) [inline flash.py]:
```suggestion
        raise Exception(output.returncode)
```

File: flash.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+import os
+import subprocess
+import re
+
+def downloadESPtool():
+    output = subprocess.run(["sudo", "apt", "install", "esptool"], text=True)
+    if(output.returncode != 0):
```
Comment (AlexErAmp) [inline flash.py]: I think best practice doing like this:
```suggestion
try:
    subprocess.run(["sudo", "apt", "install", "esptool"], check=True)
except subprocess.CalledProcessError as err:
   raise err

```

File: flash.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+def getNumber(maxInput):
+    num = 0
+    while True:
+        try:
+            num = int(input())
+            if (num >= 0 and num < maxInput):
+                break;
+            else: 
+                raise Exception("Wrong input number")
+        except:
```
Comment (AlexErAmp) [inline flash.py]: Don't use bare except. What if I want to exit at this location with ctrl-c?

File: flash.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    if(output.returncode != 0):
+        raise Exeption(output.returncode)
+        
+def getNumber(maxInput):
+    num = 0
+    while True:
+        try:
+            num = int(input())
+            if (num >= 0 and num < maxInput):
+                break;
```
Comment (AlexErAmp) [inline flash.py]: `;` - not needed

File: flash.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    output = subprocess.run(["sudo", "apt", "install", "esptool"], text=True)
+    if(output.returncode != 0):
+        raise Exeption(output.returncode)
+        
+def getNumber(maxInput):
+    num = 0
+    while True:
+        try:
+            num = int(input())
+            if (num >= 0 and num < maxInput):
```
Comment (AlexErAmp) [inline flash.py]: No parentheses needed.

File: flash.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            else: 
+                raise Exception("Wrong input number")
+        except:
+            print("Print a correct number.")
+    return num
+    
+def getCOMPortList():
+    ports = os.listdir("/dev/")
+    output = []
+    for i in range(len(ports) - 1):
```
Comment (AlexErAmp) [inline flash.py]: More pythonic:
```suggestion
    for port in ports:
```

File: flash.py (hunk 1 above)
Comment (AlexErAmp) [inline flash.py]: 1) No parentheses needed.
2) Comparison to None should be 'if cond is not None:'

File: flash.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    
+def getCOMPort():
+    print()
+    ports = getCOMPortList()
+    while (len(ports) == 0):
+        print("You have not any COM ports. Plug in your Espressif device. Then press ENTER.")
+        input()
+        ports = getCOMPorts()
+    print("available COM ports:")
+    for i in range(len(ports)):
```
Comment (AlexErAmp) [inline flash.py]:
```suggestion
    for port in ports:
```

File: flash.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        except:
+            print("Print a correct number.")
+    return num
+    
+def getCOMPortList():
+    ports = os.listdir("/dev/")
+    output = []
+    for i in range(len(ports) - 1):
+        if((re.search("ttyUSB*", ports[i]) != None) or (re.search("ttyACM*", ports[i]) != None)):
+            output.append(ports[i]);
```
Comment (AlexErAmp) [inline flash.py]: `;` - not needed

File: flash.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+
+
+try:
+    downloadESPtool()
+    COMPort = getCOMPort()
+    firmware = "./firmwares/" + getFirmware() + "/" + getESPFlashSize(COMPort) + ".bin"
+    flash(COMPort, firmware)
+except Exception as e:
+    print(e);
```
Comment (AlexErAmp) [inline flash.py]: `;` - not needed

File: flash.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    
+def getFirmware():
+    print()
+    firmwares = getFirmwareList()
+    while (len(firmwares) == 0):
+        print("Error: firmares are lost. Reload this folder from http://wiki.amperka.ru/")
+        input()
+        exit()
+    print("available firmwares:")
+    for i in range(len(firmwares)):
```
Comment (AlexErAmp) [inline flash.py]:
```suggestion
    for firmware in firmwares:
```

**Review by AlexErAmp — CHANGES_REQUESTED** [review]

**Review by nkrkv — COMMENTED** [review]
Please run source through [Black](https://pypi.org/project/black/) and keep PEP 8 naming scheme

File: flash.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        stdout=subprocess.PIPE,
+        text=True,
+        check=True
+    )
+    return re.search(".MB", out.stdout).group(0)
+
+
+def flash(com_port, firmware_path):
+    """flash board with esptool"""
+    subprocess.run(
```
Comment (AlexErAmp) [inline flash.py]: If `check` is `true`, and the process exits with a non-zero exit code, a `CalledProcessError` exception will be raised. You need to handle this exception.

File: flash.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    print("available firmwares:")
+    for i, item in enumerate(firmwares, 0):
+        print(str(i) + " " + item)
+    print("Choose your firmware. Print number of your firmware.")
+    return firmwares[get_number(len(firmwares))]
+
+
+def get_esp_flash_size(com_port):
+    """get esp8266 flash size for correct firmware"""
+    out = subprocess.run(
```
Comment (AlexErAmp) [inline flash.py]: If `check` is `true`, and the process exits with a non-zero exit code, a `CalledProcessError` exception will be raised. You need to handle this exception.

File: flash.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    """flash board with esptool"""
+    subprocess.run(
+        ["esptool", "--port", "/dev/" + com_port, "write_flash", "0x0000", firmware_path],
+        text=True,
+        check=True
+    )
+
+
+
+if not check_esptool():
```
Comment (AlexErAmp) [inline flash.py]: Place below code in `main()` function. Call the `main` function from the construct:
`if __name__ == "__main__":`

**Review by AlexErAmp — CHANGES_REQUESTED** [review]

File: flash.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+            check=True
+        )
+    except subprocess.CalledProcessError as error:
+        print("")
+        if error.returncode == 2:
+            print("Error: firmares are lost. Reload this folder from your source.")
+        elif error.returncode == 1:
+            print("Error: connection lost. Reconnect your board and relaunch script.")
+        elif error.returncode == 5:
+            print("Error: connection lost. Reconnect your board and relaunch script.")
```
Comment (AlexErAmp) [inline flash.py:91]: Duplicate code. You can do this:
```suggestion
elif error.returncode == 1 or error.returncode == 5:
     print("Error: connection lost. Reconnect your board and relaunch script.")
```

File: flash.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+        "/dev/" + com_port,
+        "write_flash",
+        "0x0000",
+        firmware_path], True)
+
+
+
+def main():
+    """main function"""
+    if not check_esptool():
```
Comment (AlexErAmp) [inline flash.py]: Add a regular exit via ctrl-c via `try: ... except KeyboardInterrupt`

File: flash.py [hunk 17]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python3
+"""script for easy update esp8266 firmware with esptool"""
```
Comment (AlexErAmp) [inline flash.py]: From PEP 257 -- Docstring Conventions:
The docstring of a script (a stand-alone program) should be usable as its "usage" message, printed when the script is invoked with incorrect or missing arguments (or perhaps with a "-h" option, for "help"). Such a docstring should document the script's function and command line syntax, environment variables, and files. Usage messages can be fairly elaborate (several screens full) and should be sufficient for a new user to use the command properly, as well as a complete quick reference to all options and arguments for the sophisticated user.

Please, add informatin about usage. Also add information that the script works with python version higher than 3.7.

**Review by AlexErAmp — CHANGES_REQUESTED** [review]

File: flash.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+        return subprocess.run(
+            command,
+            stdout=subprocess.PIPE,
+            text=True,
+            check=True
+        )
+    except subprocess.CalledProcessError as error:
+        print("")
+        if error.returncode == 2:
+            print("Error: firmares are lost. Reload this folder from your source.")
```
Comment (nkrkv) [inline flash.py]: Cannot understand the message. Do you mean something like:

> The `./firmwares` directory not found. The working directory should be a clone of https://github.com/amperka/ESP8266-uploader-script

**Review by nkrkv — APPROVED** [review]
Looks good.

Minor issues below and few notes on terms:

- Linux has directories, not folders
- Linux has serial port devices, not COM ports

**Review by AlexErAmp — APPROVED** [review]
LGTM :+1:

**Review by nkrkv — APPROVED** [review]
:heart:

**Pull request merged**